[SPARK-7328] [MLlib] [PySpark] Pyspark.mllib.linalg.Vectors: Missing items

MechCoder opened the pull request:
Add
1. Class methods squared_dist
3. parse
4. norm
5. numNonzeros
6. copy

I made a few vectorizations wrt squared_dist and dot as well. I have added support for SparseMatrix serialization in a separate PR (https://github.com/apache/spark/pull/5775) and plan to complete support for Matrices in another PR.

MechCoder committed:
[SPARK-7328] Pyspark.mllib.linalg.Vectors: Missing items

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

MechCoder commented:
ping @jkbradley @mengxr Please have a look!

SparkQA commented:
[Test build #31699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31699/consoleFull) for   PR 5872 at commit [`686b0de`](https://github.com/apache/spark/commit/686b0de6fbf249b369ff030178e491d3e2c418b4).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31701 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31701/consoleFull) for   PR 5872 at commit [`350aec7`](https://github.com/apache/spark/commit/350aec7f1d65e942e7bb7f67939dc4fb45be7bb4).

SparkQA commented:
[Test build #31701 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31701/consoleFull) for   PR 5872 at commit [`350aec7`](https://github.com/apache/spark/commit/350aec7f1d65e942e7bb7f67939dc4fb45be7bb4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31701/
Test FAILed.

SparkQA commented:
[Test build #31699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31699/consoleFull) for   PR 5872 at commit [`686b0de`](https://github.com/apache/spark/commit/686b0de6fbf249b369ff030178e491d3e2c418b4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31699/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31707 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31707/consoleFull) for   PR 5872 at commit [`59909f9`](https://github.com/apache/spark/commit/59909f9ec97b59db53ba7f7fc6d4ea1e0059609a).

SparkQA commented:
[Test build #31707 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31707/consoleFull) for   PR 5872 at commit [`59909f9`](https://github.com/apache/spark/commit/59909f9ec97b59db53ba7f7fc6d4ea1e0059609a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31707/
Test PASSed.

MechCoder commented:
@mengxr Sorry for spaming but could you have a look at this too? My exams start this Friday and hence would like to complete the linalg module at least by then.

jkbradley commented:
I'll take a look now

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -208,9 +209,55 @@ def __init__(self, ar):
+        6.0
+        """
+        if p == 1:
+            return np.sum(np.abs(self.array))
```
You can use ```numpy.linalg.norm(self.array, p)```

MechCoder replied:
oops

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -208,9 +209,55 @@ def __init__(self, ar):
         return DenseVector, (self.array.tostring(),)
 
+    def numNonzeros(self):
+        return np.nonzero(self.array)[0].size
```
You can use np.count_nonzero

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -208,9 +209,55 @@ def __init__(self, ar):
+        >>> DenseVector.parse('[0.0,1.0,2.0,3.0]')
+        DenseVector([0.0, 1.0, 2.0, 3.0])
+        """
+        vectorString = vectorString[1:-1]
```
It'd be nice if the "[" and "]" were found more robustly, perhaps using string.find.

MechCoder replied:
I see. I thought this was a standard form and parse should return results only for those :p

jkbradley replied:
I think the Scala version is a bit more robust since it uses Java's StringTokenizer for splitting

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -208,9 +209,55 @@ def __init__(self, ar):
+        """
+        return str(self)
+
+    def copy(self):
```
I would guess we don't need this either.  But I'll ping @mengxr --- How does this differ from ```copy.deepcopy()```?

MechCoder replied:
it should do the same thing.

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -208,9 +209,55 @@ def __init__(self, ar):
             ar = ar.astype(np.float64)
         self.array = ar
 
+    def toString(self):
```
This is not necessary.  In Python, we use __str__, and that's already implemented

jkbradley commented:
I just accidentally sent some comments, but I'm not done yet

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -387,9 +434,60 @@ def __init__(self, size, *args):
                 if self.indices[i] >= self.indices[i + 1]:
                     raise TypeError("indices array must be sorted")
 
+    def copy(self):
```
My comments about DenseVector methods carry over to copy + other SparseVector methods

jkbradley commented:
Could you please put the vectorization in another JIRA & PR?  It's great to have, but separation is nice for both reviewing and debugging if any issues show up after the merge.

jkbradley commented on python/pyspark/mllib/tests.py:
```diff
@@ -217,6 +221,27 @@ def test_dense_matrix_is_transposed(self):
         self.assertTrue(array_equal(sm.colPtrs, [0, 2, 5]))
         self.assertTrue(array_equal(sm.values, [1, 3, 4, 6, 9]))
 
+    def test_parse_matrix(self):
```
should be named "test_parse_vector"

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -387,9 +434,60 @@ def __init__(self, size, *args):
+        >>> SparseVector.parse('(4,[0,1],[4.0,5.0])')
+        SparseVector(4, {0: 4.0, 1: 5.0})
+        """
+        size = int(vectorString[1])
```
This should be more robust to spaces, etc.

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -633,6 +743,57 @@ def stringify(vector):
         return str(vector)
 
+    @staticmethod
+    def dot(a, b):
```
Is this needed?  It's not in Vectors.scala

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -633,6 +743,57 @@ def stringify(vector):
+        return a.dot(b)
+
+    @staticmethod
+    def squared_distance(a, b):
```
There's an inconsistency here:
* Python: ```DenseVector.squared_distance```, ```SparseVector.squared_distance```
* Scala: ```Vectors.sqdist```

I'm OK with naming this squared_distance or sqdist, or with having both with one as an alias.  CC @mengxr  Votes?

MechCoder replied:
I'm thinking we should keep both for a couple of versions, deprecate squared_distance and advice users to use sqdist. (At least that's how it works in sklearn.)

jkbradley replied:
Sounds good.  Let's go for "sqdist."

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -633,6 +743,57 @@ def stringify(vector):
+    def squared_distance(a, b):
+        """
+        Squared distance between two vectors.
+        a and b can be of type, SparseVector, DenseVector, np.ndarray
```
"type," --> "type" (no comma)

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -633,6 +743,57 @@ def stringify(vector):
+
+    @staticmethod
+    def parse(vectorString):
+        if vectorString[0] == '[':
```
It'd be nice for this to be more robust, as in Scala

jkbradley commented:
That's all for now!

MechCoder committed:
Robust parser and removed unnecessary methods

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31991 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31991/consoleFull) for   PR 5872 at commit [`1bd3c04`](https://github.com/apache/spark/commit/1bd3c0498596927feca961eb23ff4cc018230889).

MechCoder commented:
@jkbradley I've addressed all your comments. Let me know if the parse function is now robust enough. I've updated the docstring example to show that it indeed is.

SparkQA commented:
[Test build #31991 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31991/consoleFull) for   PR 5872 at commit [`1bd3c04`](https://github.com/apache/spark/commit/1bd3c0498596927feca961eb23ff4cc018230889).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31991/
Test PASSed.

MechCoder commented:
Also https://github.com/apache/spark/pull/5946

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -208,9 +208,37 @@ def __init__(self, ar):
         self.array = ar
 
+    @staticmethod
+    def parse(vectorString):
```
Can the argument please be "s" to match Scala?  I just thought of this issue.

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -387,8 +415,50 @@ def __init__(self, size, *args):
+            (self.size, self.indices.tostring(), self.values.tostring()))
+
+    @staticmethod
+    def parse(vectorString):
```
parse() needs to print out error messages upon failed parses; please see the Scala version, which prints the full line which failed to parse.

Can the argument please be "s" to match Scala?

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -387,8 +415,50 @@ def __init__(self, size, *args):
+        start = vectorString.find('(')
+        end = vectorString.find(')')
+        vectorString = vectorString[start+1: end].strip()
+        size = int(vectorString[0])
```
This only looks at 1 character, which won't work for size > 9

MechCoder replied:
Great. I'll add a test for this.

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -633,6 +703,38 @@ def stringify(vector):
         return str(vector)
 
+    @staticmethod
+    def squared_distance(a, b):
```
Can the arguments please be "v1" and "v2" to match Scala?

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -633,6 +703,38 @@ def stringify(vector):
+        return a.squared_distance(b)
+
+    @staticmethod
+    def norm(vec, p):
```
Can the argument please be "vector" instead of "vec" to match Scala?

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -633,6 +703,38 @@ def stringify(vector):
+        return _convert_to_vector(vec).norm(p)
+
+    @staticmethod
+    def parse(vectorString):
```
Can you please add doc, plus little doc tests?

Also, can the argument please be "s" to match Scala?

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -633,6 +703,38 @@ def stringify(vector):
+        return SparseVector.parse(vectorString)
+
+    @staticmethod
+    def zeros(num):
```
Can the argument please be "size" to match Scala?

MechCoder committed:
Add error message for parser

MechCoder commented:
@jkbradley updated!

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32080 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32080/consoleFull) for   PR 5872 at commit [`3a9da1e`](https://github.com/apache/spark/commit/3a9da1ede69041f690c7a1a0dde95c738930b907).

SparkQA commented:
[Test build #32080 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32080/consoleFull) for   PR 5872 at commit [`3a9da1e`](https://github.com/apache/spark/commit/3a9da1ede69041f690c7a1a0dde95c738930b907).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32080/
Test FAILed.

jkbradley commented:
Jenkins test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32088/consoleFull) for   PR 5872 at commit [`3a9da1e`](https://github.com/apache/spark/commit/3a9da1ede69041f690c7a1a0dde95c738930b907).

SparkQA commented:
[Test build #32088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32088/consoleFull) for   PR 5872 at commit [`3a9da1e`](https://github.com/apache/spark/commit/3a9da1ede69041f690c7a1a0dde95c738930b907).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32088/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32096 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32096/consoleFull) for   PR 5872 at commit [`0b7bd69`](https://github.com/apache/spark/commit/0b7bd69ef0a63bad0f775fb0c3b56e055053a84a).

SparkQA commented:
[Test build #32096 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32096/consoleFull) for   PR 5872 at commit [`0b7bd69`](https://github.com/apache/spark/commit/0b7bd69ef0a63bad0f775fb0c3b56e055053a84a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32096/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32104 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32104/consoleFull) for   PR 5872 at commit [`ce3e53e`](https://github.com/apache/spark/commit/ce3e53edb967758f36a77a244d2a0a7152618fc4).

SparkQA commented:
[Test build #32104 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32104/consoleFull) for   PR 5872 at commit [`ce3e53e`](https://github.com/apache/spark/commit/ce3e53edb967758f36a77a244d2a0a7152618fc4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32104/
Test FAILed.

MechCoder commented:
test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32108 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32108/consoleFull) for   PR 5872 at commit [`ce3e53e`](https://github.com/apache/spark/commit/ce3e53edb967758f36a77a244d2a0a7152618fc4).

SparkQA commented:
[Test build #32108 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32108/consoleFull) for   PR 5872 at commit [`ce3e53e`](https://github.com/apache/spark/commit/ce3e53edb967758f36a77a244d2a0a7152618fc4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32108/
Test PASSed.

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -208,9 +208,46 @@ def __init__(self, ar):
+            raise ValueError("Array should start with '['")
+        end = s.find(']')
+        if end == -1:
+            raise ValueError("Array should end with  ']")
```
missing single quote

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -208,9 +208,46 @@ def __init__(self, ar):
+        try:
+            values = [float(val) for val in s.split(',')]
+        except ValueError:
+            raise ValueError("Unable to parse values.")
```
Please print full string s here

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -387,8 +424,73 @@ def __init__(self, size, *args):
+
+        ind_start = s.find('[')
+        if ind_start == -1:
+            raise ValueError("Indices array should start with '('.")
```
should be bracket, not parenthesis (same below)

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -387,8 +424,73 @@ def __init__(self, size, *args):
+        try:
+            indices = [int(ind) for ind in ind_list]
+        except ValueError:
+            raise ValueError("Unabel to parse indices.")
```
typo: "Unable"
Also, please print full string s

jkbradley commented on python/pyspark/mllib/linalg.py:
```diff
@@ -387,8 +424,73 @@ def __init__(self, size, *args):
+        try:
+            values = [float(val) for val in val_list]
+        except ValueError:
+            raise ValueError("Unable to parse values.")
```
Print full string s

jkbradley commented:
OK, I think that's it.

MechCoder committed:
minor

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32133/consoleFull) for   PR 5872 at commit [`9bd269d`](https://github.com/apache/spark/commit/9bd269dc1e917388f2cebd45b90d98205088657f).

MechCoder commented:
@jkbradley fixed!

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32135/consoleFull) for   PR 5872 at commit [`a8ff1e0`](https://github.com/apache/spark/commit/a8ff1e09bdc0baec7d10acdbb2cf789c121af7dc).

SparkQA commented:
[Test build #32135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32135/consoleFull) for   PR 5872 at commit [`a8ff1e0`](https://github.com/apache/spark/commit/a8ff1e09bdc0baec7d10acdbb2cf789c121af7dc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32135/
Test FAILed.

MechCoder commented:
test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32143 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32143/consoleFull) for   PR 5872 at commit [`a8ff1e0`](https://github.com/apache/spark/commit/a8ff1e09bdc0baec7d10acdbb2cf789c121af7dc).

SparkQA commented:
[Test build #32133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32133/consoleFull) for   PR 5872 at commit [`9bd269d`](https://github.com/apache/spark/commit/9bd269dc1e917388f2cebd45b90d98205088657f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32133/
Test PASSed.

jkbradley commented:
LGTM Merging into master and branch-1.4   Thanks!

MechCoder commented:
Thanks! I have my exams from 8th to 18th. So will be away for a while, trying to study for those.. Just a heads up!

jkbradley commented:
OK, good luck with them!

SparkQA commented:
**[Test build #32143 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32143/consoleFull)**     for PR 5872 at commit [`a8ff1e0`](https://github.com/apache/spark/commit/a8ff1e09bdc0baec7d10acdbb2cf789c121af7dc)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32143/
Test FAILed.
